File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        root = TreeNode(preorder[0])
        inOrderIndex = inorder.index(preorder[0])
        leftLength = len(inorder[:inOrderIndex])
        rightLength = len(inorder[inOrderIndex + 1:])

        root.left = self.buildTree(preorder[1:1 + inOrderIndex], inorder[:inOrderIndex])
        root.right = self.buildTree(preorder[1 + inOrderIndex:], inorder[1 + inOrderIndex:])

        return root

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        List<Integer> list = new ArrayList<>();
        inorder(root, list, k);
        return list.get(k - 1);
    }

    private void inorder(TreeNode node, List<Integer> list, int k) {
        if (node == null || list.size() == k) {
            return;
        }

        inorder(node.left, list, k);
        list.add(node.val);
        inorder(node.right, list, k);

    }
}
